packages/core/src/components/forms/numericInput.tsx

```diff
@@ -351,11 +375,33 @@ export class NumericInput extends AbstractComponent<HTMLInputProps & INumericInp
             this.incrementValue(delta);
         }
 
+        if (this.props.allowNumericCharactersOnly && this.isKeyboardEventDisabledForBasicNumericEntry(e)) {
+            e.preventDefault();
+        }
+
         Utils.safeInvoke(this.props.onKeyDown, e);
     }
 
+    private handleInputPaste = (e: React.ClipboardEvent<HTMLInputElement>) => {
+        this.didPasteEventJustOccur = true;
+        Utils.safeInvoke(this.props.onPaste, e);
+    }
+
     private handleInputChange = (e: React.FormEvent<HTMLInputElement>) => {
-        const nextValue = (e.target as HTMLInputElement).value;
+        const value = (e.target as HTMLInputElement).value;
+
+        let nextValue: string;
+
+        if (this.props.allowNumericCharactersOnly && this.didPasteEventJustOccur) {
+            this.didPasteEventJustOccur = false;
```

Comment:
Yeah it's pretty obvious what's going on, i was only curious because you mentioned it's one of the inelegant place. A comment sounds good.